src/Resolver.hack

```diff
@@ -113,7 +113,16 @@ final class Resolver {
                 throw new \Error('Unsupported operation: '.$operation_type);
         }
 
-        return tuple($result->getValue(), $result->getErrors());
+        while ($result is DeferredFieldResult<_>) {
+            // This is where we'd batch load data for all pending promises
+            $result = await $result->resolveAsync();
+        }
+
+        if ($result is ValidFieldResult<_>) {
+            return tuple($result->getValue(), $result->getErrors());
+        } else {
```

Comment:
I unfortunately lost some type safety when I changed the return type of `NullableOutputType::resolveAsync` from `ValidFieldResult<?T>` to `FieldResult<?T>`. However, now that `NullableOutputType::resolveAsync` can return a deferred result, I wasn't sure how to get around this.